pkg/services/hcloud/server/server.go

```diff
@@ -90,62 +98,52 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
 
 	conditions.MarkTrue(s.scope.HCloudMachine, infrav1.BootstrapReadyCondition)
 
-	// try to find an existing server
-	server, err := s.findServer(ctx)
-	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to get server: %w", err)
+	switch s.scope.HCloudMachine.Status.BootState {
+	case infrav1.HCloudBootStateUnset:
+		return s.handleBootStateUnset(ctx)
+	case infrav1.HCloudBootStateBootToRealOS:
+		return s.handleBootToRealOS(ctx)
+	default:
+		return reconcile.Result{}, fmt.Errorf("unknown BootState: %s", s.scope.HCloudMachine.Status.BootState)
 	}
+}
 
-	// if no server is found we have to create one
-	if server == nil {
-		server, err = s.createServer(ctx)
-		if err != nil {
-			if errors.Is(err, errServerCreateNotPossible) {
-				return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
-			}
-			return reconcile.Result{}, fmt.Errorf("failed to create server: %w", err)
+func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, error) {
+	m := s.scope.HCloudMachine
+	if m.Spec.ProviderID != nil && *m.Spec.ProviderID != "" {
+		// This machine seems to be an existing machine which was created before introducing
+		// ImageURL.
+
+		if !m.Status.Ready {
```

Comment:
you could set the state on the one you use in the new code as well (handleBootToRealOS I believe)

Reply:
@janiskemper at the end of handleBootToRealOS there is `s.scope.SetReady(true)`. Okay like this?

Reply:
ok, now I understood you. new code:

```go 
	if m.Spec.ProviderID != nil && *m.Spec.ProviderID != "" {
		// This machine seems to be an existing machine which was created before introducing
		// ImageURL.

		if !m.Status.Ready {
			m.SetBootState(infrav1.HCloudBootStateBootToPreRescueOS)
			return reconcile.Result{RequeueAfter: requeueAfterCreateServer}, nil
		}
		m.SetBootState(infrav1.HCloudBootStateOperatingSystemRunning)
		return reconcile.Result{}, nil
	}
```

Reply:
yess